bioimageio/spec/utils/build_spec.py

```diff
@@ -0,0 +1,246 @@
+        kwargs['preprocessing'] = preprocessing
+
+    inputs = raw_nodes.InputTensor(
+        name='input' if name is None else name,
+        data_type=str(test_in.dtype),
+        axes='bczyx' if test_in.ndim == 5 else 'bcyx',
+        shape=shape_description,
+        **kwargs
+    )
+    return inputs
+
+
+def _get_output_tensor(test_out, name,
+                       reference_input, scale, offset,
+                       postprocessing, halo):
+    shape = test_out.shape
+    if reference_input is None:
+        assert scale is None
+        assert offset is None
+        shape_description = shape
+    else:
+        assert scale is not None
+        assert offset is not None
+        shape_description = {
+            'reference_input': reference_input,
+            'scale': scale,
+            'offset': offset
+        }
+
+    kwargs = {}
+    if postprocessing is not None:
+        kwargs['postprocessing'] = postprocessing
+    if halo is not None:
+        kwargs['halo'] = halo
+
+    outputs = raw_nodes.OutputTensor(
+        name='output' if name is None else name,
+        data_type=str(test_out.dtype),
+        axes='bczyx' if test_out.ndim == 5 else 'bcyx',
+        shape=shape_description,
+        **kwargs
+    )
+    return outputs
+
+
+# TODO can we pattern match to decide if we have a url or doi?
+def _build_cite(cite):
+    citation_list = [
+        raw_nodes.CiteEntry(text=k, url=v) for k, v in cite.items()
+    ]
+    return citation_list
+
+
+# NOTE does not support multiple input / output tensors yet
+# to implement this we should wait for 0.4.0, see also
+# https://github.com/bioimage-io/spec-bioimage-io/issues/70#issuecomment-825737433
+def build_spec(
+    # model specific required
+    source: str,
+    model_kwargs: Dict[str, Union[int, float, str]],
+    weight_uri: str,
+    test_inputs: List[str],
+    test_outputs: List[str],
+    # general required
+    name: str,
+    description: str,
+    authors: List[str],
+    tags: List[str],
+    license: str,
+    documentation: str,
+    covers: str,
+    dependencies: str,
+    cite: Optional[List[str]],
+    root: Optional[str] = None,
+    # model specific optional
+    weight_type: Optional[str] = None,
+    sample_inputs: Optional[str] = None,
+    sample_outputs: Optional[str] = None,
+    # tensor specific
+    input_name: Optional[str] = None,
+    input_step: Optional[List[int]] = None,
+    input_min_shape: Optional[List[int]] = None,
+    output_name: Optional[str] = None,
+    output_reference: Optional[str] = None,
+    output_scale: Optional[List[int]] = None,
+    output_offset: Optional[List[int]] = None,
+    halo: Optional[List[int]] = None,
+    preprocessing: Optional[List[Dict[str, Dict[str, Union[int, float, str]]]]] = None,
+    postprocessing: Optional[List[Dict[str, Dict[str, Union[int, float, str]]]]] = None,
+    # general optional
+    git_repo: Optional[str] = None,
+    attachments: Optional[List[str]] = None,
```

Comment:
spec-bioimage-io: 
> Dictionary of text keys...

```suggestion
    attachments: Optional[Dict[str, Any]] = None,
```

Reply:
To be completely precise it says:

> attachments [optional] Dictionary of text keys and URI (or a list of URI) 

So both dict and list. But I am not really sure why we would need a dict here; a flat list of files / URIs should be enough.

Reply:
oh, I read that as a dict of text keys and [URI (or a list of URI)]...
and that's what I have implemented in the validator
```python
    attachments = fields.Dict(
        fields.String,
        fields.Union([fields.URI(), fields.List(fields.URI)]),
        ...
        ```
I think the idea was to give the URIs some explicit name/tag/mini documentation and @oeway had suggested some special keys, e.g. "files" (or smth similar) to be downloaded and included by the packager...

Reply:
Ok, that makes sense :). Then it should be `Optional[Dict[str, Union[str, List[str]]]` though.